kernel/src/scan/log_replay.rs

```diff
@@ -37,80 +37,108 @@ struct LogReplayScanner {
-struct AddRemoveVisitor {
-    adds: Vec<(Add, usize)>,
-    removes: Vec<Remove>,
-    selection_vector: Option<Vec<bool>>,
-    // whether or not we are visiting commit json (=true) or checkpoint (=false)
+struct AddRemoveDedupVisitor<'seen> {
+    seen: &'seen mut HashSet<FileActionKey>,
+    selection_vector: Vec<bool>,
     is_log_batch: bool,
 }
+impl<'seen> AddRemoveDedupVisitor<'seen> {
+    fn already_seen(&mut self, path: &str, dv_unique_id: Option<String>) -> bool {
+        // Note: each (add.path + add.dv_unique_id()) pair has a
+        // unique Add + Remove pair in the log. For example:
+        // https://github.com/delta-io/delta/blob/master/spark/src/test/resources/delta/table-with-dv-large/_delta_log/00000000000000000001.json
 
-const ADD_FIELD_COUNT: usize = 15;
-
-impl AddRemoveVisitor {
-    fn new(selection_vector: Option<Vec<bool>>, is_log_batch: bool) -> Self {
-        AddRemoveVisitor {
-            selection_vector,
-            is_log_batch,
-            ..Default::default()
+        let key = FileActionKey::new(path, dv_unique_id);
+        if self.seen.contains(&key) {
+            debug!(
+                "Ignoring duplicate ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            true
+        } else {
+            debug!(
+                "Including ({}, {:?}) in scan, is log {}",
+                key.path, key.dv_unique_id, self.is_log_batch
+            );
+            if self.is_log_batch {
+                // Remember file actions from this batch so we can ignore duplicates as we process
+                // batches from older commit and/or checkpoint files. We don't track checkpoint
+                // batches because they are already the oldest actions and never replace anything.
+                self.seen.insert(key);
+            }
+            false
         }
     }
-}
 
-impl RowVisitorBase for AddRemoveVisitor {
+    fn filter_row<'a>(&mut self, i: usize, getters: &[&'a dyn GetData<'a>]) -> DeltaResult<bool> {
+        // Add will have a path at index 0 if it is valid; otherwise, if it is a log batch, we
+        // may have a remove with a path at index 4.
+        let (path, getters, is_add) = if let Some(path) = getters[0].get_opt(i, "add.path")? {
+            (path, &getters[1..4], true)
+        } else if !self.is_log_batch {
+            return Ok(false);
+        } else if let Some(path) = getters[4].get_opt(i, "remove.path")? {
+            (path, &getters[5..8], false)
+        } else {
+            return Ok(false);
+        };
+
+        let dv_unique_id = match getters[0].get_opt(i, "deletionVector.storageType")? {
+            Some(storage_type) => Some(DeletionVectorDescriptor::unique_id_from_parts(
+                storage_type,
+                getters[1].get(i, "deletionVector.pathOrInlineDv")?,
+                getters[2].get_opt(i, "deletionVector.offset")?,
+            )),
+            None => None,
```

Comment:
could just do a `.map(...)`?

Reply:
A `map` call would intercept those `?`, which is annoying and bulky to deal with:
```rust
        let dv_unique_id = getters[0]
            .get_opt(i, "deletionVector.storageType")?
            .map(|storage_type| {
                DeletionVectorDescriptor::unique_id_from_parts(
                    storage_type,
                    getters[1].get(i, "deletionVector.pathOrInlineDv")?,
                    getters[2].get_opt(i, "deletionVector.offset")?,
                )
            })
            .transpose()?;
```

Reply:
oh I was thinking of a `storage_type.map(DeletionVectorDescriptor...)` but probably has the same `?` issue?